base/reduce.jl

```diff
@@ -280,29 +280,65 @@ mapreduce_impl(f, op, A::AbstractArrayOrBroadcasted, ifirst::Integer, ilast::Int
 """
     mapreduce(f, op, itrs...; [init])
 
-Apply function `f` to each element(s) in `itrs`, and then reduce the result using the binary
-function `op`. If provided, `init` must be a neutral element for `op` that will be returned
-for empty collections. It is unspecified whether `init` is used for non-empty collections.
-In general, it will be necessary to provide `init` to work with empty collections.
+Apply function `f` to each element(s) in `itrs`, and then repeatedly call the 2 argument
+function `op` with those results or results from previous `op` evaluations until a single value is returned.
+
+If provided, `init` is the return value for empty collections and is used one or more times as
+an argument to `op` for non-empty collections. The `init` value is not transformed by the function `f`.
+Using `init` ensures that all calls to `op` have one argument that is either `init`
+or the result from a previous `op` evaluation, and the ordering of these arguments is
+unspecified. As it may appear in the reduction one or more times, it must be a neutral element for `op` that
+does not change the result by being used more than once. It is generally an error to call `mapreduce`
+with empty collections without specifying an `init` value, but in unambiguous cases an
+identity value for `op` may be returned; see [`Base.reduce_empty`](@ref) for more details.
+
+In contrast with [`mapfoldl`](@ref) and [`mapfoldr`](@ref), the sequence of
+function evaluations and the associativity of the reduction is not specified
+and may vary between different methods and across Julia versions. Some implementations
+may reuse the return value of `f` for elements that appear multiple times in the
+collection(s).
```

Comment:
I'm uneasy with the idea that `f`'s return values may be re-used rather than repeatedly calling `f`. 

It's very normal to have an impure `f` function even for a parallel `mapreduce` so long as calls to `f` don't depend on the order of operations. For example, it's totally normal for people to do  `I/O` in `f`, or they might want to return a mutable collection, where it'd be incorrect and buggy to re-use the same collection.

For example, a parallelizable `map!(f, dest, coll)` can be pretty efficiently implemented as follows:
```julia
function mymap!(f, dest, coll)
    mapreduce((l, r) -> nothing, eachindex(dest, coll)) do i
        dest[i] = f(coll[i])
    end
end
```

Reply:
This is not a new part of the docstring, though.  It dates way back to #10536.

Reply:
Maybe clarify that this normally only happens for "sparse" data structures that implicitly store some repeated element?

Reply:
Hm I see. That's a bit scary.

Reply:
After some thought, I like @stevengj's suggestion. I think it's a sensible corner case to make it clear to people that we're not normally in the buisness of memoizing reductions, but if a datastructure has its own memoization scheme then we'll reserve the right to also take advantage of that.

Reply:
I'd ask that this be tackled separately — I fiddled around with talking about memoized/sparse/something, but it's so hard to talk about in the abstract.  I've opened issue #55238 to not lose the thread.